buildreport/buildreport.go

```diff
@@ -0,0 +1,454 @@
+				reports: r,
+			}
+		}
+	}
+	// Either the BEGIN/END markers couldn't be found, or DATA couldn't be found. Keep before and
+	// after (if found), but ignore the content of the data section (if any).
+	return commentBody{
+		before: before,
+		after:  after,
+	}
+}
+
+func (c *commentBody) update(report State) {
+	found := false
+	for i := range c.reports {
+		r := &c.reports[i]
+		if r.ID == report.ID {
+			// Update the found report.
+			r.updateFrom(report)
+			found = true
+			break
+		}
+	}
+	if !found {
+		c.reports = append(c.reports, report)
+	}
+}
+
+func (c *commentBody) body() (string, error) {
+	var b strings.Builder
+	b.WriteString(c.before)
+	// We can properly parse a comment when its text runs directly into the data markers, but for
+	// ease of reading the generated Markdown source manually, insert newlines. Make sure there are
+	// two, so it looks the same as a paragraph break.
+	if c.before != "" && !strings.HasSuffix(c.before, "\n") {
+		b.WriteString("\n")
+		if !strings.HasSuffix(c.before, "\n\n") {
+			b.WriteString("\n")
+		}
+	}
+	b.WriteString(beginDataSectionMarker)
+	b.WriteString("\n")
+
+	sort.SliceStable(c.reports, func(i, j int) bool {
+		iv, jv := c.reports[i], c.reports[j]
+		if c := strings.Compare(iv.Version, jv.Version); c != 0 {
+			return c < 0
+		}
+		if c := strings.Compare(iv.Name, jv.Name); c != 0 {
+			return c < 0
+		}
+		if c := strings.Compare(iv.ID, jv.ID); c != 0 {
+			return c < 0
+		}
+		return false
+	})
+
+	var version, name string
+	// Always start a new table for the first report, even if it has no version or no name.
+	newTable := true
+	for _, r := range c.reports {
+		if r.Version != version {
+			version = r.Version
+			b.WriteString("\n## ")
+			b.WriteString(version)
+			b.WriteString("\n")
+			name = ""
+			newTable = true
+		}
+		if r.Name != name {
+			name = r.Name
+			b.WriteString("\n### ")
+			b.WriteString(name)
+			b.WriteString("\n\n")
+			newTable = true
+		}
+		if newTable {
+			newTable = false
+			b.WriteString("| ID | Status | Started | Last Report |\n")
+			b.WriteString("| --- | :---: | --- | --- |\n")
+		}
+		b.WriteString("| ")
+		if r.URL != "" {
+			b.WriteString("[")
+		}
+		b.WriteString(r.ID)
+		if r.URL != "" {
+			b.WriteString("](")
+			b.WriteString(r.URL)
+			b.WriteString(")")
+			// If the build has failed (potentially needs retry) and is a release infra build that
+			// publishes detailed retry information on the "Extensions" tab, then show a direct
+			// link.
+			if r.Status == SymbolFailed {
+				if _, ok := pipelinesWithRetryInstructions[r.Name]; ok {
+					b.WriteString(" ([Retry](")
+					b.WriteString(r.URL)
+					b.WriteString("&view=ms.vss-build-web.run-extensions-tab))")
+				}
+			}
+		}
+		b.WriteString(" | ")
+		b.WriteString(r.Status)
+		b.WriteString(" | ")
+		if r.StartTime != (time.Time{}) {
+			b.WriteString(r.StartTime.Format("2006-01-02 15:04 MST"))
+		}
+		b.WriteString(" | ")
+		if r.LastUpdate != (time.Time{}) {
```

Comment:
The idiomatic way to do this check is using [time.IsZero()](https://pkg.go.dev/time#Time.IsZero).